### DISCUSSION TITLE
Custom method instrumentation when running .java file "directly"

### DISCUSSION THREAD

**Discussion by breedx-splk** [open-telemetry/opentelemetry-java-instrumentation#7820] [Q&A]
I figured this was way too oddball/niche to bring up as an issue, but in doing some testing work today I discovered (the hard way) that the `OTEL_INSTRUMENTATION_METHODS_INCLUDE` setting that leverages the `io.opentelemetry.methods` tracer does not seem to be compatible with running `.java` files directly.

So in my case, something like this:

```
OTEL_INSTRUMENTATION_METHODS_INCLUDE="MyCoolClass[myCoolMethod]" \
java -javaagent:opentelemetry-javaagent.jar MyCoolClass.java
```

was failing to actually instrument the method. But if I did an explicit `javac` and instead run the resulting `.class` things work ok.

Anyway, I assume that there is a lifecycle or race condition with the classloaders, and that a screwball path happens when running `java Whatever.java` directly. 

I would not expect many users to be impacted by this, but wanted this to be here for posterity. If folks think this is interesting or have insight, feel free to share.

**Comment by trask** [comment]
> running `java Whatever.java` directly

what is this magic? 🛸

**Reply by breedx-splk** [reply]
I know, right!? It was supposed to be a shortcut....it was instead a find-the-problem side quest.

Anyway, I guess that came along in Java 11? I suspect nobody uses it outside of demos/talks. 🙃

**Comment by laurit** [comment]
Actually this seems to work on jdk17 and fail on jdk11. This is because there is a class resource presence check in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/8f32133430a78455cce418252a413f8b2f78f93e/instrumentation/methods/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/methods/MethodInstrumentation.java#L38-L40 but the class loader used to load this class does not implement resource finders in jdk11 https://github.com/openjdk/jdk11u-dev/blob/d7a6e6ae495a780c6818dd69ae240ac092e41504/src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/Main.java#L533 but does in jdk17 https://github.com/openjdk/jdk/blob/46bcc4901ea66589ec4c6904f8bab9156422312a/src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/Main.java#L571

**Reply by breedx-splk** [reply]
Oh interesting for sure. I didn't try jdk17 in this case.

**Reply by trask** [reply]
wow, nice find
